rpmlb/builder/koji.py

```diff
@@ -0,0 +1,173 @@
+"""Builder interface for the Koji build service."""
+
+import re
+from pathlib import Path
+from typing import Iterator, Optional, Sequence
+
+from ..utils import run_cmd, run_cmd_with_capture
+from ..work import Work
+from .base import BaseBuilder
+
+
+class KojiBuilder(BaseBuilder):
+    """Builder interface for the Koji build service."""
+
+    DEFAULT_TARGET_FORMAT = 'sclo{epel}-{collection}-rh-el{epel}'
+
+    def __init__(self):
+        """Initialize the slots"""
+
+        self.collection = None  #: The collection to build
+        self.epel = None  #: Version of EPEL to build for
+        self.profile = None  #: The koji configuration profile to use
+        #: The format string for construction of the actual template name
+        self.target_format = type(self).DEFAULT_TARGET_FORMAT
+
+    @property
+    def base_command(self) -> Sequence[str]:
+        """Basis of the command for build service interaction."""
+
+        binary = 'koji'
+
+        if self.profile is None:
+            return [binary]
+        else:
+            return [binary, '--profile', self.profile]
+
+    @property
+    def target(self) -> str:
+        """The koji target to build into."""
+
+        return self.target_format.format(
+            collection=self.collection,
+            epel=self.epel,
+        )
+
+    def run(self, work: Work, **kwargs):
```

Comment:
Builder sub class should not override `run` function due to `is-a` design.
Builder class is also intended as a template method design pattern [1].
Super class defines the behavior with high abstraction level. and the sub classes only follow the behavior.

[1] https://en.wikipedia.org/wiki/Template_method_pattern